keps/sig-node/5526-pod-level-resource-managers/README.md

```diff
@@ -0,0 +1,1800 @@
+    across uncore cache groups, enhancing shared cache locality for the overall
+    pod level CPU pool.
+
+#### Pod Scope Allocation and Partitioning Algorithm
+
+The following algorithm describes the process for the CPU Manager:
+
+1.  **[New Logic] One-Time Pod-Level Calculation:** When the `Allocate` function
+    is called for the first container of a pod, the manager performs a one-time
+    calculation for the entire pod. It determines the total pod resource budget
+    from `pod.spec.resources` and uses the Topology Manager's hint to identify
+    the single, NUMA-aligned (if a suitable alignment was found) set of CPUs
+    that will be managed for the pod.
+
+2.  **[New Logic] Calculation of Shared Pool:** The manager calculates the
+    `podSharedPool` by taking the total allocated pool and subtracting the sum
+    of all exclusive CPU requests from the pod's `Guaranteed` containers (both
+    init and app containers).
+
+    A fundamental concept in this new functionality is the `podSharedPool`. This
+    represents the set of resources within the pod's total NUMA-aligned budget
+    that are available to be shared by all containers that do not have their own
+    exclusive `Guaranteed` resource requests. This mechanism is key to the
+    partitioning model, as it ensures that critical containers receive their
+    reserved resources while allowing all other containers to efficiently share
+    the remainder of the pod's budget.
+
+3.  **Handling Init Containers:**
+
+    -   **Standard Init Containers:** For init containers that are individually
+        `Guaranteed`, an exclusive slice is carved out from the pool. Upon
+        completion, these CPUs are added to a per-pod `reusable` pool, making
+        them available to subsequent app containers within the pool.
+    -   **Restartable Init Containers (Sidecars):** For restartable init
+        containers that are individually `Guaranteed`, an exclusive slice is
+        carved out and reserved for the entire pod lifecycle. These CPUs are
+        *not* added to the reusable pool.
+
+4.  **[New Logic] Pre-computation and State Update:** The manager pre-computes
+    and writes the specific assignment for every container to the state file.
+    `Guaranteed` containers are assigned their exclusive slices, and all other
+    containers are assigned the `podSharedPool`.
+
+5.  **Container-Level Allocation:** For the current container (and all
+    subsequent containers in the pod), the `Allocate` function reads that
+    container's specific, pre-computed assignment from the state and applies it
+    to the container's cgroup.
+
+#### State Management and Container Removal
+
+The in-memory state and on-disk checkpoint file will be updated to be pod-aware.
+For the `pod` scope, the state will now store the total managed resource set for
+the pod and the individual assignments for each container (either an exclusive
+slice or the `podSharedPool`).
```

Comment:
this is fine and most likely necessary, but we need to handle it carefully in the upgrade flow and in the upgrade-downgrade (revert) flow .